ptx/src/pass/emit_llvm.rs

```diff
@@ -2281,15 +2282,51 @@ impl<'a> MethodEmitContext<'a> {
     ) -> Result<(), TranslateError> {
         let src1 = self.resolver.value(arguments.src1)?;
         let src2 = self.resolver.value(arguments.src2)?;
-        self.emit_intrinsic(
-            c"llvm.amdgcn.mul.u24",
+        let name_lo = match data.type_ {
+            ast::ScalarType::U32 => c"llvm.amdgcn.mul.u24",
+            ast::ScalarType::S32 => c"llvm.amdgcn.mul.i24",
+            _ => return Err(error_unreachable()),
+        };
+        let res_lo = self.emit_intrinsic(
+            name_lo,
             Some(arguments.dst),

```

Comment:
This should be `None` if it is the `.hi` variant. `Some(x)` means "emit intrinsic and register return `LLVMValueRef` to `x`", `None` means "emit instrinsic and return `LLVMValueRef`, the returned value does not correspond to any ZLUDA identifier - it's some intermediate LLVM value".
It all works right now somewhat by accident because in `.hi` case we override the value down below, but strictly speaking goes against invariants.